From 7b576262fbf07e9314b651a9e575125fd35ed942 Mon Sep 17 00:00:00 2001
From: Tomas Henzl <thenzl@redhat.com>
Date: Mon, 8 Sep 2014 14:32:03 -0400
Subject: [scsi] hpsa: report check condition even if no sense data present for ioaccel2 mode

Message-id: <1410186734-19986-73-git-send-email-thenzl@redhat.com>
Patchwork-id: 92081
O-Subject: [RHEL6.6 e-stor PATCH 72/83] hpsa: report check condition even if no sense data present for ioaccel2 mode
Bugzilla: 1069185
RH-Acked-by: Maurizio Lombardi <mlombard@redhat.com>
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Ewan Milne <emilne@redhat.com>

It shouldn't happen that we get a check condition with no sense data, but if it
does, we shouldn't just drop the check condition on the floor.

Signed-off-by: Stephen M. Cameron <scameron@beardog.cce.hp.com>
Reviewed-by: Justin Lindley <justin.lindley@hp.com>
Reviewed-by: Mike Miller <michael.miller@canonical.com>
Signed-off-by: Christoph Hellwig <hch@lst.de>
(cherry picked from commit ee6b18890fb680cf4f9036d238ee7eea582421fd)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/hpsa.c b/drivers/scsi/hpsa.c
index 906e882..9722de8 100644
--- a/drivers/scsi/hpsa.c
+++ b/drivers/scsi/hpsa.c
@@ -1562,9 +1562,13 @@ static int handle_ioaccel_mode2_error(struct ctlr_info *h,
    dev_warn(&h->pdev->dev,
     "%s: task complete with check condition.\n",
     "HP SSD Smart Path");
+   cmd->result |= SAM_STAT_CHECK_CONDITION;
    if (c2->error_data.data_present !=
-     IOACCEL2_SENSE_DATA_PRESENT)
+     IOACCEL2_SENSE_DATA_PRESENT) {
+    memset(cmd->sense_buffer, 0,
+     SCSI_SENSE_BUFFERSIZE);
     break;
+   }
    /* copy the sense data */
    data_len = c2->error_data.sense_data_len;
    if (data_len > SCSI_SENSE_BUFFERSIZE)
@@ -1574,7 +1578,6 @@ static int handle_ioaccel_mode2_error(struct ctlr_info *h,
      sizeof(c2->error_data.sense_data_buff);
    memcpy(cmd->sense_buffer,
     c2->error_data.sense_data_buff, data_len);
-   cmd->result |= SAM_STAT_CHECK_CONDITION;
    retry = 1;
    break;
   case IOACCEL2_STATUS_SR_TASK_COMP_BUSY:
-- 
1.7.1